Mathlib/GroupTheory/DivisibleHull2.lean

```diff
@@ -0,0 +1,419 @@
+  le_refl a := by
+    induction a using ind with | mk m s
+    rw [mk_le_mk]
+  le_trans a b c hab hbc := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    rw [mk_le_mk, Submonoid.smul_def] at ⊢ hab hbc
+    rw [← (nsmul_right_strictMono (show sb.val ≠ 0 by simp)).le_iff_le]
+    rw [← (nsmul_right_strictMono (show sc.val ≠ 0 by simp)).le_iff_le] at hab
+    rw [← (nsmul_right_strictMono (show sa.val ≠ 0 by simp)).le_iff_le] at hbc
+    rw [smul_comm _ _ ma, smul_comm _ _ mc]
+    rw [smul_comm _ _ mb] at hab
+    exact hab.trans hbc
+  le_antisymm a b h h' := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    rw [mk_le_mk] at h h'
+    rw [mk_eq_mk_iff_smul_eq_smul]
+    exact le_antisymm h h'
+  le_total a b := by
+    induction a using ind with | mk ma sa
+    induction b using ind with | mk mb sb
+    exact le_total _ _
+  toDecidableLE := decidable_liftOn₂ _ _
+
+theorem mk_lt_mk {m m' : M} {s s' : nonZeroDivisors ℕ} : mk m s < mk m' s' ↔ s' • m < s • m' := by
+  simp_rw [lt_iff_not_ge, mk_le_mk]
+
+theorem add_lt_add_left (a : DivisibleHull M) {b c : DivisibleHull M} (h : b < c) :
+    a + b < a + c := by
+  induction a using ind with | mk ma sa
+  induction b using ind with | mk mb sb
+  induction c using ind with | mk mc sc
+  simp_rw [mk_add_mk]
+  rw [mk_lt_mk] at ⊢ h
+  simp_rw [mul_smul]
+  simp_rw [Submonoid.smul_def] at ⊢ h
+  simp_rw [smul_add, smul_smul]
+  suffices (sa * sb * sc : ℕ) • ma + (sa * sa : ℕ) • (sc : ℕ) • mb
+      < (sa * sb * sc : ℕ) • ma + (sa * sa : ℕ) • (sb : ℕ) • mc by
+    simp_rw [smul_smul] at this
+    convert this using 3 <;> ring
+  apply _root_.add_lt_add_left <| nsmul_lt_nsmul_right (by simp) h
+
+instance : IsOrderedCancelAddMonoid (DivisibleHull M) where
+  add_le_add_left a b h c := by
+    obtain rfl | h := eq_or_lt_of_le h
+    · simp
+    exact (add_lt_add_left c h).le
+  le_of_add_le_add_left a b c h := by
+    contrapose! h
+    exact add_lt_add_left a h
+
+instance : PosSMulStrictMono ℚ≥0 (DivisibleHull M) where
+  smul_lt_smul_of_pos_left a ha b c h := by
+    induction b using ind with | mk mb sb
+    induction c using ind with | mk mc sc
+    simp_rw [mk_lt_mk, Submonoid.smul_def] at h
+    simp_rw [nnqsmul_mk, mk_lt_mk, Submonoid.smul_def, Submonoid.mul_def, smul_smul]
+    simp_rw [mul_right_comm _ _ a.num, mul_smul _ _ mc, mul_smul _ _ mb]
+    refine (nsmul_right_strictMono ?_).lt_iff_lt.mpr h
+    simpa using ha.ne.symm
+
+variable {M : Type*} [AddCommGroup M] [LinearOrder M] [IsOrderedAddMonoid M]
+
+instance : PosSMulStrictMono ℚ (DivisibleHull M) where
+  smul_lt_smul_of_pos_left a ha b c h := by
+    simpa only [qsmul_def, ha.le, ↓reduceDIte] using smul_lt_smul_of_pos_left h (by simpa using ha)
+
+variable (M) in
+/-- The map `m ↦ m / 1` as an `OrderAddMonoidHom`. -/
+@[simps]
+def mkOrderAddMonoidHom : M →+o DivisibleHull M where
```

Comment:
I think this should be a coercion (or rather, you should have a coe instance defined in terms of this). Perhaps you could even start writing `n⁻¹ • ↑x` instead of `mk x n`?